src/devices/src/legacy/serial.rs

```diff
@@ -276,34 +288,72 @@ impl BusDevice for Serial {
 }
 
 impl Subscriber for Serial {
-    /// Handle a read event (EPOLLIN) on the serial input fd.
-    fn process(&mut self, event: &EpollEvent, _: &mut EventManager) {
+    /// Handle events on the serial input fd.
+    fn process(&mut self, event: &EpollEvent, ev_mgr: &mut EventManager) {
         let source = event.fd();
         let event_set = event.event_set();
 
-        // TODO: also check for errors. Pending high level discussions on how we want
-        // to handle errors in devices.
-        let supported_events = EventSet::IN;
-        if !supported_events.contains(event_set) {
```

Comment:
The problem seems to be limited to this line.

The `epoll_ctl` man page says:
```
EPOLLHUP
              Hang up happened on the associated file descriptor.  epoll_wait(2) will always wait for this event; it is not necessary to set it in
              events.
```

Meaning we should always be prepared to handle `EventSet::HANG_UP`.

I believe just removing this error shortcircuit here by adding `EventSet::HANG_UP` to `supported_events` will fix the issue.

**FYI: This problem exists for all our devices, so it should be fixed everywhere**, also documentation for `EventManager` should be updated to clearly specify that `EventSet::HANG_UP` can be triggered even if it wasn't declared in `interest_list()`.

Reply:
Do we already have an issue for the other devices as well?

Similarly, for edge triggered events we need to handle EPOLLRDHUP.

Reply:
> I believe just removing this error shortcircuit here by adding EventSet::HANG_UP to supported_events will fix the issue.

I've tested this shortcircuit removal and to me it seems that it isn't enough. `EventSet::HANG_UP` it is level-triggered and keeps coming again and again, even if we handle all the inflight bytes.

This keeps the emulation thread at 100% (due to the continuously event manager dispatching of the event to the serial). The solution from this PR for this is to unregister the stdin fd and basically dropping later on any serial input altogether.

As far as my testing goes, the way one process can use *pipes* is by using *anonymous pipes* or *named pipes*.

If one process uses anonymous pipes (through `int pipe(int pipefd[2])` or similar `pipe2` call) and write end of the pipe is closed, then the read end will continuously receive `HANG_UP`, because it can not communicate anymore with the closed end. This event can not be fixed and the communication is broken/should terminate.

However, if a process uses named pipes (through `int mkfifo(const char *pathname, mode_t mode)` it behaves different. 

Let's open a read end and a write end on top of this named pipe:
```
let rc = unsafe { libc::mkfifo(path.as_ptr(), libc::O_NONBLOCK as u32) };
assert!(rc == 0);

let mut fds: [c_int; 2] = [0; 2];

// Open as an non-blocking read end pipe, to be able to work on a single thread.
fds[0] = unsafe { libc::open(path.as_ptr(), libc::O_RDONLY | libc::O_NONBLOCK) };
assert!(fds[0] >= 0);

fds[1] = unsafe { libc::open(path.as_ptr(), libc::O_WRONLY) };
assert!(fds[1] >= 0);
```

Now, let's close the write end. This will start flooding the read end with `HANG_UP`s. However, we can stop these events by re-opening a file on top of the named pipe with `O_WRONLY` flag, which basically fixes the write end of the pipe. We can send now bytes on this pipe as if nothing happened. However, a simple bash example somehow contradicts this practice:
```
# Terminal 1 - read end.
mkfifo fifo
# Blocking fifo.
cat < fifo
# Terminal 2 - write end.
cat > fifo &
export write_pid=$!
sudo gdb -p  $write_pid
> print close(1) # Close process standard input (pipe write end).
```

This makes the process reading from the pipe to end. This is the default behavior for `cat` at least, since `HANG_UP` is seen as an end-of-file.

My thoughts about this is that it is mandatory to not keep the emulation thread at 100% utilization because of flooding it with `HANG_UP`s. At the same time, it would be great if we could support scenarios of named pipes, like the one described above, even if they seem a hack/workaround for something I can't think of right now. At the moment, I can not see a solution without compromises and if we have to choose, I think the 100% issue is prioritary.

Do you guys see any scenario where unregistering the stdin fd from the event manager (and stopping the HANG_UP) might impair any use case which tries to re establish the pipe "connection" after closing the write end? Should we really take it into consideration for a real world scenario?

Reply:
> Do we already have an issue for the other devices as well?

We have an issue for this: #1627.

> Similarly, for edge triggered events we need to handle EPOLLRDHUP.

It is already mentioned in the issue above.

Reply:
> It is already mentioned in the issue above.

Cool, thanks!

Reply:
> My thoughts about this is that it is mandatory to not keep the emulation thread at 100% utilization because of flooding it with HANG_UPs. At the same time, it would be great if we could support scenarios of named pipes.

We shouldn't worry about the named pipe scenario. We can work with the limitation that once the STDIN pipe is closed, it is irrecoverable.

With this info brought to light by @iulianbarbu's investigation, I am fine with unregistering the FDs on `HANG_UP`.

Reply:
But in this case, I am thinking whether it'd be better to do that in `EventManager` instead of repeating same code for all `Subscriber`s.
@andreeaflorescu what do you think?

Reply:
> But in this case, I am thinking whether it'd be better to do that in `EventManager` instead of repeating same code for all `Subscriber`s.

I think there might be cases that depend on `HANG_UP` handling outside the event manager loop, inside the application thread. Wouldn't this drop the support for such cases?

I do not have a real example, at the moment, but because pipes can be used as IPC, it might replace some behavior obtained through unix signaling. In case of `SIGHUP`, some daemons might [reload their configuration](https://stackoverflow.com/questions/19052354/sighup-for-reloading-configuration).

This implies custom handling that needs to be done and if the event manager assumes a default behavior for `HANG_UP`, this custom handling wouldn't be possible, right?

Reply:
I wouldn't add this in the EventManager. I find it pretty hard to assume a correct, desired, default behavior for `HANG_UP`. We could make sure to not duplicate the code everywhere by providing some Firecracker specific handlers, that can be used in all devices implementation.

For example, we can define an `EventHandler<T> where T: EventSubscriber` and have methods on it for `handle_error()`, `handle_rhup()` and so on. This way we can reduce the code duplication. I didn't put much thought into this, I am sure we can find some other better alternatives.